torch/nn/modules/activation.py

```diff
@@ -1117,21 +1115,41 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
             elif torch.is_grad_enabled() and any([x.requires_grad for x in tensor_args]):
                 why_not_fast_path = ("grad is enabled and at least one of query or the "
                                      "input/output projection weights or biases requires_grad")
+
             if not why_not_fast_path:
-                return torch._native_multi_head_attention(
-                    query,
-                    key,
-                    value,
-                    self.embed_dim,
-                    self.num_heads,
-                    self.in_proj_weight,
-                    self.in_proj_bias,
-                    self.out_proj.weight,
-                    self.out_proj.bias,
-                    key_padding_mask if key_padding_mask is not None else attn_mask,
-                    need_weights,
-                    average_attn_weights,
-                    1 if key_padding_mask is not None else 0 if attn_mask is not None else None)
+                if self.batch_first:
+                    return torch._native_multi_head_attention(
+                        query,
+                        key,
+                        value,
+                        self.embed_dim,
+                        self.num_heads,
+                        self.in_proj_weight,
+                        self.in_proj_bias,
+                        self.out_proj.weight,
+                        self.out_proj.bias,
+                        key_padding_mask if key_padding_mask is not None else attn_mask,
+                        need_weights,
+                        average_attn_weights,
+                        1 if key_padding_mask is not None else 0 if attn_mask is not None else None)
+                else:  # If batch_first is not true we need to transpose the input
+                    # for fast-path non self-attention is used (query, key value are same tensors)
+                    query = key = value = query.transpose(1, 0)
```

Comment:
You can have this line and the one that transpose back in a `if not self.batch_first` so that you don't have to duplicate the rather verbose call to `_native_multi_head_attention`.

Reply:
The reason why I chose to duplicate the verbose call to `_native_multi_head_attention` is for performance reasons since it would do only one check for `self.batch_first` instead of two. Since performance is key, I chose this approach over code conciseness.

Reply:
Did you confirm this makes a difference in performance? I highly, highly doubt it. `self.batch_first` is a bool.

Reply:
No it did not make a difference.